# CAO Points Analysis

[CAO](http://www.cao.ie/index.php?page=points&p=2021)

***

#### Importing Libaries

In [1]:
# Data Frames
import pandas as pd

from collections import namedtuple

# Read in PDF
import requests, PyPDF2

# Regular Expressions
import re

# Convenient HTTP Requests.
import requests as rq

# dates and times
import datetime as dt

# For downloading
import urllib.request as urlrq

# For Comparison
import datacompy

# Numerical Arrays
import numpy as np

# Cooler plots
import seaborn as sns

import matplotlib.pyplot as plt

#### Date and Time for saving data

In [2]:
# Get Current date and time.
now = dt.datetime.now()

# Format as string
nowstr = now.strftime('%Y%m%d_%H%M%S')

# 2021 Points

#### Save original data set

In [3]:
# Fetch CAO points URL
resp = rq.get('http://www2.cao.ie/points/l8.php')
# Have a quick peek
resp

<Response [200]>

In [4]:
#Create file path for original data
pathhtml = 'data/cao2021_'+ nowstr + '.html'

##### Error on Server
- Server says we should decode as per:
    Content-Type: text/html; charset=iso-8859-1
- However one line uses \x96 which is not defined in iso-8859-1
- Therefore we used the similar decoding standard cp1252
- This is very similar but includes \x96

In [5]:
# Server uses the incorrect encoding
original_encoding = resp.encoding
# Change to cp1252
resp.encoding = 'cp1252'

In [6]:
# Save The orginal html file
with open(pathhtml, 'w') as f:
    f.write(resp.text)

In [7]:
# Quick Check to confirm encoding
resp.encoding

'cp1252'

# Use regular Expressions to select the lines we want

In [8]:
# Compile Reg-Ex for Matching lines

re_course = re.compile(r'([A-Z]{2}[0-9]{3})(.*)')

```def points_to_array(s):
    portfolio = ''
    if s[0] == '#':
        portfolio = '#'
    random = ''
    if s[-1] == '*':
        random = 'True'
    points = ''
    for i in s:
        if i.isdigit():
            points = points + i
    return [points, portfolio, random]```


In [9]:
# File path for csv file
path2021 = 'data/cao2021_csv_'+ nowstr + '.csv'

In [10]:

# Keep track of how many courses we process.
no_lines = 0

# Open CSV file for writing 
with open (path2021, 'w') as f:
    # Write a header row
    f.write(','.join(['Code', 'Title', 'PointsRD1', 'PointsRD2']) + '\n')
    # loop through lines of the response
    for line in resp.iter_lines():
        #decode the line using 'cp1252' - wrong encoding
        dline = line.decode('cp1252')
        # Match only the lines representing courses.
        if re_course.fullmatch(dline):
            # Add one to the lines counter.
            no_lines = no_lines + 1
            # The course code
            course_code = dline[:5]
            #print (course_code)
            # course title
            course_title = dline[7:57].strip()
            #print (course_title)
            # Round one points
            course_points = re.split(' +', dline[60:])
            if len(course_points) !=2:
                course_points = course_points[ :2]
            # join the fields using a comma
            linesplit = [course_code, course_title, course_points[0], course_points[1]]
            # Rejoin the substrings with commas
            f.write(','.join(linesplit) + '\n')

# Print the total number of processed lines.
print(f"Total number of lines is {no_lines}.")

Total number of lines is 949.


<br>

**NB:** it was verified as of 09/11/2021 that there were 949 courses exactly in the cao2021 points list

In [11]:
df2021 = pd.read_csv(path2021)

In [12]:
df2021

Code                                              Title PointsRD1  \
0    AL801     Software Design for Virtual Reality and Gaming       300   
1    AL802  Software Design in Artificial Intelligence for...       313   
2    AL803  Software Design for Mobile Apps and Connected ...       350   
3    AL805    Computer Engineering for Network Infrastructure       321   
4    AL810                                 Quantity Surveying       328   
..     ...                                                ...       ...   
944  WD211                                 Creative Computing       270   
945  WD212                    Recreation and Sport Management       262   
946  WD230           Mechanical and Manufacturing Engineering       230   
947  WD231                 Early Childhood Care and Education       266   
948  WD232                       Business Information Systems       261   

    PointsRD2  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
..        ...  
944       NaN  
945       NaN  
946       230  
947       NaN  
948       261  

[949 rows x 4 columns]

In [13]:
df2021['PointsRD1'] = df2021['PointsRD1'].str.replace(r'\W',"")

<ipython-input-13-469e21142b27>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df2021['PointsRD1'] = df2021['PointsRD1'].str.replace(r'\W',"")


In [14]:
df2021

Code                                              Title PointsRD1  \
0    AL801     Software Design for Virtual Reality and Gaming       300   
1    AL802  Software Design in Artificial Intelligence for...       313   
2    AL803  Software Design for Mobile Apps and Connected ...       350   
3    AL805    Computer Engineering for Network Infrastructure       321   
4    AL810                                 Quantity Surveying       328   
..     ...                                                ...       ...   
944  WD211                                 Creative Computing       270   
945  WD212                    Recreation and Sport Management       262   
946  WD230           Mechanical and Manufacturing Engineering       230   
947  WD231                 Early Childhood Care and Education       266   
948  WD232                       Business Information Systems       261   

    PointsRD2  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
..        ...  
944       NaN  
945       NaN  
946       230  
947       NaN  
948       261  

[949 rows x 4 columns]

<br>

# 2020 CAO Points

In [15]:
url2020 = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

In [16]:
#Create file path for original data
pathxlsx = 'data/cao2020_'+ nowstr + '.xlsx'

In [17]:
# Save original file to disk
urlrq.urlretrieve(url2020, pathxlsx)

('data/cao2020_20211228_231757.xlsx',
 <http.client.HTTPMessage at 0x214837de0d0>)

<br> 

## Load Spreadsheet using Pandas

In [18]:
# Download and parse the excel spreadsheet
df2020 = pd.read_excel('http://www2.cao.ie/points/CAOPointsCharts2020.xlsx', skiprows = 10)

In [19]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [20]:
# Spot Check Random Row
df2020.iloc[753]

CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Road Transport Technology and Management
COURSE CODE2                                                           LC286
R1 POINTS                                                                264
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      264
EOS Random *                                                             NaN
EOS Mid-point                                                            360
LEVEL                                                                      7
HEI                                         Limerick Institute of Technology
Test/Interview #                                                         NaN

In [21]:
# Spot Check End of Data
df2020.iloc[-1]

CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Mechanical and Manufacturing Engineering
COURSE CODE2                                                           WD230
R1 POINTS                                                                253
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      253
EOS Random *                                                             NaN
EOS Mid-point                                                            369
LEVEL                                                                      8
HEI                                        Waterford Institute of Technology
Test/Interview #                                                         NaN

In [22]:
#Create file path for PANDAS data
path = 'data/cao2020_'+ nowstr + '.csv'

In [23]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

When cleaning the data here I could not seem to get rid of the some of the strings and special characters without corrupting the data in the 'R1 POINTS' column. I tried lots of different ways but the data kept returning with the string gone but the points totals I wanted gone too, or there was simply no change to other columns. 
The ways i tried included :
 - ```.str.replace('#+matric', '')```  
 - ```.map(lambda x: x.lstrip('+-').rstrip('aAbBcC'))```  
 - ```df[cols_to_check] = df[cols_to_check].replace({';':''}, regex=True)```
 
 I spent a huge amount of time here tryingto resolve this and had no sucess. I was extremly frustrated at this point and time was speeding 

In [25]:
# Save pandas data frame to disk
df2020.to_csv(path)

## 2019 Points following lecture Video

http://www2.cao.ie/points/lvl8_19.pdf

##### Steps to reproduce

 - Download orginal PDF
 - Open Original pdf in Microsoft Word
 - Save Microsoft Words Coverted PDF
 - Re save version in word for editing
 - Delete Headers and footers
 - Delete intro on page 1
 - Select All and copy
 - Paste into notepad++
 - Remove HEI name headings and paste into each course line
 - Delete Blank Lines
 - Replace Double Tab Characters with Single Tab
 - Deleted all tab characters at end 
 - Replace all backticks with commas

In [26]:
df2019 = pd.read_csv('data/2019POINTS_20211104103000_edited.csv', sep= '\t')

In [27]:
df2019

Course Code                                             Course  EOS    Mid
0         AL801    Software Design with Virtual Reality and Gaming  304  328.0
1         AL802               Software Design with Cloud Computing  301  306.0
2         AL803  Software Design with Mobile Apps and Connected...  309  337.0
3         AL805        Network Management and Cloud Infrastructure  329  442.0
4         AL810                                 Quantity Surveying  307  349.0
..          ...                                                ...  ...    ...
925       WD200                                     Arts (options)  221  296.0
926       WD210                       Software Systems Development  271  329.0
927       WD211                                 Creative Computing  275  322.0
928       WD212                    Recreation and Sport Management  274  311.0
929       WD230           Mechanical and Manufacturing Engineering  273  348.0

[930 rows x 4 columns]

In [28]:
# Removes Special Characters
df2019['EOS'] = df2019['EOS'].str.replace(r'\W',"")

<ipython-input-28-06bc442524b1>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  df2019['EOS'] = df2019['EOS'].str.replace(r'\W',"")


In [29]:
# Create a file path for the pandas data.
path2019 = 'data/cao2019_' + nowstr + '.csv'

In [30]:
# Save pandas data frame to disk.
df2019.to_csv(path2019)

# Quick look at the 3 data frames

In [31]:

df2021

Code                                              Title PointsRD1  \
0    AL801     Software Design for Virtual Reality and Gaming       300   
1    AL802  Software Design in Artificial Intelligence for...       313   
2    AL803  Software Design for Mobile Apps and Connected ...       350   
3    AL805    Computer Engineering for Network Infrastructure       321   
4    AL810                                 Quantity Surveying       328   
..     ...                                                ...       ...   
944  WD211                                 Creative Computing       270   
945  WD212                    Recreation and Sport Management       262   
946  WD230           Mechanical and Manufacturing Engineering       230   
947  WD231                 Early Childhood Care and Education       266   
948  WD232                       Business Information Systems       261   

    PointsRD2  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
..        ...  
944       NaN  
945       NaN  
946       230  
947       NaN  
948       261  

[949 rows x 4 columns]

In [32]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [33]:
df2019

Course Code                                             Course  EOS    Mid
0         AL801    Software Design with Virtual Reality and Gaming  304  328.0
1         AL802               Software Design with Cloud Computing  301  306.0
2         AL803  Software Design with Mobile Apps and Connected...  309  337.0
3         AL805        Network Management and Cloud Infrastructure  329  442.0
4         AL810                                 Quantity Surveying  307  349.0
..          ...                                                ...  ...    ...
925       WD200                                     Arts (options)  221  296.0
926       WD210                       Software Systems Development  271  329.0
927       WD211                                 Creative Computing  275  322.0
928       WD212                    Recreation and Sport Management  274  311.0
929       WD230           Mechanical and Manufacturing Engineering  273  348.0

[930 rows x 4 columns]

<br>

## Concatonate and Join

<br>

Concatonate and Join

In [34]:
courses2021 = df2021[['Code', 'Title']]
courses2021

Code                                              Title
0    AL801     Software Design for Virtual Reality and Gaming
1    AL802  Software Design in Artificial Intelligence for...
2    AL803  Software Design for Mobile Apps and Connected ...
3    AL805    Computer Engineering for Network Infrastructure
4    AL810                                 Quantity Surveying
..     ...                                                ...
944  WD211                                 Creative Computing
945  WD212                    Recreation and Sport Management
946  WD230           Mechanical and Manufacturing Engineering
947  WD231                 Early Childhood Care and Education
948  WD232                       Business Information Systems

[949 rows x 2 columns]

In [35]:
courses2020 = df2020[['COURSE CODE2','COURSE TITLE']]
courses2020.columns = ['Code', 'Title']
courses2020

Code                                              Title
0     AC120                             International Business
1     AC137                                       Liberal Arts
2     AD101   First Year Art & Design (Common Entry,portfolio)
3     AD102  Graphic Design and Moving Image Design (portfo...
4     AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
1459  WD208                          Manufacturing Engineering
1460  WD210                       Software Systems Development
1461  WD211                                 Creative Computing
1462  WD212                    Recreation and Sport Management
1463  WD230           Mechanical and Manufacturing Engineering

[1464 rows x 2 columns]

In [36]:
courses2019 = df2019[['Course Code','Course']]
courses2019.columns = ['Code', 'Title']
courses2019

Code                                              Title
0    AL801    Software Design with Virtual Reality and Gaming
1    AL802               Software Design with Cloud Computing
2    AL803  Software Design with Mobile Apps and Connected...
3    AL805        Network Management and Cloud Infrastructure
4    AL810                                 Quantity Surveying
..     ...                                                ...
925  WD200                                     Arts (options)
926  WD210                       Software Systems Development
927  WD211                                 Creative Computing
928  WD212                    Recreation and Sport Management
929  WD230           Mechanical and Manufacturing Engineering

[930 rows x 2 columns]

In [37]:
allcourses = pd.concat([courses2021, courses2020, courses2019], ignore_index=True)
allcourses

Code                                              Title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[3343 rows x 2 columns]

In [38]:
allcourses.sort_values('Code')

Code                                     Title
175   AC120                    International Business
949   AC120                    International Business
2581  AC120                    International Business
950   AC137                              Liberal Arts
2582  AC137                              Liberal Arts
...     ...                                       ...
2412  WD230  Mechanical and Manufacturing Engineering
946   WD230  Mechanical and Manufacturing Engineering
3342  WD230  Mechanical and Manufacturing Engineering
947   WD231        Early Childhood Care and Education
948   WD232              Business Information Systems

[3343 rows x 2 columns]

In [39]:
# finds all extra copies of duplicated rolls
allcourses[allcourses.duplicated()]

Code                                              Title
949   AC120                             International Business
950   AC137                                       Liberal Arts
952   AD102  Graphic Design and Moving Image Design (portfo...
955   AD204                               Fine Art (portfolio)
956   AD211                         Fashion Design (portfolio)
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[1405 rows x 2 columns]

In [40]:
#Copy of the combo of 2021, 2020 and 2019 data frame with the duplicates removed
allcourses.drop_duplicates()

Code                                              Title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
3281  TL801         Animation Visual Effects and Motion Design
3282  TL802                            TV, Radio and New Media
3283  TL803                                   Music Technology
3286  TL812                       Computing with Digital Media
3324  WD148                                  Tourism Marketing

[1938 rows x 2 columns]

In [41]:
# finds all extra copies of duplicated rolls
# based only on 'Code' column
allcourses[allcourses.duplicated(subset=['Code'])]

Code                                              Title
949   AC120                             International Business
950   AC137                                       Liberal Arts
951   AD101   First Year Art & Design (Common Entry,portfolio)
952   AD102  Graphic Design and Moving Image Design (portfo...
953   AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[1692 rows x 2 columns]

In [42]:
#Copy of the combo of 2021,2020 and 2019 data frame with the duplicates removed
# based only on 'Code' column
allcourses.drop_duplicates(subset = ['Code'], inplace = True, ignore_index=True )

In [43]:
allcourses

Code                                              Title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
1646  SG441                              Environmental Science
1647  SG446                                Applied Archaeology
1648  TL803                                   Music Technology
1649  TL812                       Computing with Digital Media
1650  WD148                                  Tourism Marketing

[1651 rows x 2 columns]

<br>

### Join the Points

In [44]:
# Set the index to the code column.
df2021.set_index('Code', inplace=True)
df2021.columns = ['title', 'points_r1_2021', 'points_r2_2021']
df2021

title points_r1_2021  \
Code                                                                      
AL801     Software Design for Virtual Reality and Gaming            300   
AL802  Software Design in Artificial Intelligence for...            313   
AL803  Software Design for Mobile Apps and Connected ...            350   
AL805    Computer Engineering for Network Infrastructure            321   
AL810                                 Quantity Surveying            328   
...                                                  ...            ...   
WD211                                 Creative Computing            270   
WD212                    Recreation and Sport Management            262   
WD230           Mechanical and Manufacturing Engineering            230   
WD231                 Early Childhood Care and Education            266   
WD232                       Business Information Systems            261   

      points_r2_2021  
Code                  
AL801            NaN  
AL802            NaN  
AL803            NaN  
AL805            NaN  
AL810            NaN  
...              ...  
WD211            NaN  
WD212            NaN  
WD230            230  
WD231            NaN  
WD232            261  

[949 rows x 3 columns]

In [45]:
# Set the index to the code column.
allcourses.set_index('Code', inplace=True)

In [46]:
allcourses = allcourses.join(df2021[['points_r1_2021']])
allcourses

Title points_r1_2021
Code                                                                   
AL801     Software Design for Virtual Reality and Gaming            300
AL802  Software Design in Artificial Intelligence for...            313
AL803  Software Design for Mobile Apps and Connected ...            350
AL805    Computer Engineering for Network Infrastructure            321
AL810                                 Quantity Surveying            328
...                                                  ...            ...
SG441                              Environmental Science            NaN
SG446                                Applied Archaeology            NaN
TL803                                   Music Technology            NaN
TL812                       Computing with Digital Media            NaN
WD148                                  Tourism Marketing            NaN

[1651 rows x 2 columns]

In [47]:
df2020_r1 = df2020[['COURSE CODE2', 'R1 POINTS']]
df2020_r1.columns = ['code', 'points_r1_2020']
df2020_r1

code points_r1_2020
0     AC120            209
1     AC137            252
2     AD101       #+matric
3     AD102       #+matric
4     AD103       #+matric
...     ...            ...
1459  WD208            188
1460  WD210            279
1461  WD211            271
1462  WD212            270
1463  WD230            253

[1464 rows x 2 columns]

In [48]:
# Set the index to the code column.
df2020_r1.set_index('code', inplace=True)
df2020_r1

points_r1_2020
code                
AC120            209
AC137            252
AD101       #+matric
AD102       #+matric
AD103       #+matric
...              ...
WD208            188
WD210            279
WD211            271
WD212            270
WD230            253

[1464 rows x 1 columns]

In [49]:
# Join 2020 points to allcourses.
allcourses = allcourses.join(df2020_r1)
allcourses

Title points_r1_2021  \
Code                                                                      
AL801     Software Design for Virtual Reality and Gaming            300   
AL802  Software Design in Artificial Intelligence for...            313   
AL803  Software Design for Mobile Apps and Connected ...            350   
AL805    Computer Engineering for Network Infrastructure            321   
AL810                                 Quantity Surveying            328   
...                                                  ...            ...   
SG441                              Environmental Science            NaN   
SG446                                Applied Archaeology            NaN   
TL803                                   Music Technology            NaN   
TL812                       Computing with Digital Media            NaN   
WD148                                  Tourism Marketing            NaN   

      points_r1_2020  
Code                  
AL801            303  
AL802            332  
AL803            337  
AL805            333  
AL810            319  
...              ...  
SG441            NaN  
SG446            NaN  
TL803            NaN  
TL812            NaN  
WD148            NaN  

[1651 rows x 3 columns]

In [50]:
df2019_r1 = df2019[['Course Code', 'EOS']]
df2019_r1.columns = ['code', 'points_r1_2019']
df2019_r1

code points_r1_2019
0    AL801            304
1    AL802            301
2    AL803            309
3    AL805            329
4    AL810            307
..     ...            ...
925  WD200            221
926  WD210            271
927  WD211            275
928  WD212            274
929  WD230            273

[930 rows x 2 columns]

In [51]:
# Set the index to the code column.
df2019_r1.set_index('code', inplace=True)
df2019_r1

points_r1_2019
code                
AL801            304
AL802            301
AL803            309
AL805            329
AL810            307
...              ...
WD200            221
WD210            271
WD211            275
WD212            274
WD230            273

[930 rows x 1 columns]

In [52]:
# Join 2019 points to allcourses.
allcourses = allcourses.join(df2019_r1)
allcourses

Title points_r1_2021  \
Code                                                                      
AL801     Software Design for Virtual Reality and Gaming            300   
AL802  Software Design in Artificial Intelligence for...            313   
AL803  Software Design for Mobile Apps and Connected ...            350   
AL805    Computer Engineering for Network Infrastructure            321   
AL810                                 Quantity Surveying            328   
...                                                  ...            ...   
SG441                              Environmental Science            NaN   
SG446                                Applied Archaeology            NaN   
TL803                                   Music Technology            NaN   
TL812                       Computing with Digital Media            NaN   
WD148                                  Tourism Marketing            NaN   

      points_r1_2020 points_r1_2019  
Code                                 
AL801            303            304  
AL802            332            301  
AL803            337            309  
AL805            333            329  
AL810            319            307  
...              ...            ...  
SG441            NaN            297  
SG446            NaN            289  
TL803            NaN            264  
TL812            NaN            369  
WD148            NaN            269  

[1651 rows x 4 columns]

### Returning to 2020 points

As I now had the data in the columns I needed I still had one problem I had some spurious characters and strings in the 2020 column - after more research and frustration and trying different methods I decided to do what I normally do when I run into an IT problem at work - I plugged it out and plugged it back in again. So here I decided to export the data I had to CSV, then read it back in again as a new dataframe called 'firstrndpts' and try do my changes again. This worked perfectly so now I had all the courses for 2019, 2020 and 2021 with all the points in a clean dataframe.

In [70]:
print(list(allcourses.columns))

['Title', 'points_r1_2021', 'points_r1_2020', 'points_r1_2019']


In [71]:
# test to ensure points columns are the same types of data - Panda series
for col in list (allcourses.columns):
    print(allcourses[col])
    print('\n')
    print(type(allcourses[col]))

Code
AL801       Software Design for Virtual Reality and Gaming
AL802    Software Design in Artificial Intelligence for...
AL803    Software Design for Mobile Apps and Connected ...
AL805      Computer Engineering for Network Infrastructure
AL810                                   Quantity Surveying
                               ...                        
SG441                                Environmental Science
SG446                                  Applied Archaeology
TL803                                     Music Technology
TL812                         Computing with Digital Media
WD148                                    Tourism Marketing
Name: Title, Length: 1651, dtype: object


<class 'pandas.core.series.Series'>
Code
AL801    300
AL802    313
AL803    350
AL805    321
AL810    328
        ... 
SG441    NaN
SG446    NaN
TL803    NaN
TL812    NaN
WD148    NaN
Name: points_r1_2021, Length: 1651, dtype: object


<class 'pandas.core.series.Series'>
Code
AL801    303
AL802    332


In [ ]:
# create csv file of data frame
allcourses.to_csv('All_Courses_RD1.csv')

In [65]:
# Read back in data in new data frame
firstrndpts = pd.read_csv('All_Courses_RD1.csv')

In [73]:
# Take a peak
firstrndpts.head()

Code                                              Title points_r1_2021  \
0  AL801     Software Design for Virtual Reality and Gaming            300   
1  AL802  Software Design in Artificial Intelligence for...            313   
2  AL803  Software Design for Mobile Apps and Connected ...            350   
3  AL805    Computer Engineering for Network Infrastructure            321   
4  AL810                                 Quantity Surveying            328   

  points_r1_2020 points_r1_2019  
0            303            304  
1            332            301  
2            337            309  
3            333            329  
4            319            307

In [74]:
# Remove #+matric, # and AQA from ata frame
firstrndpts['points_r1_2020'] = firstrndpts['points_r1_2020'].str.replace('#+matric', '', regex = False)
firstrndpts['points_r1_2020'] = firstrndpts['points_r1_2020'].str.replace('#', '', regex = False)
firstrndpts['points_r1_2020'] = firstrndpts['points_r1_2020'].str.replace('AQA', '', regex = False)

In [76]:
# Export the data to a CSV file
firstrndpts.to_csv('wih.csv')

In [75]:
firstrndpts

Code                                              Title points_r1_2021  \
0     AL801     Software Design for Virtual Reality and Gaming            300   
1     AL802  Software Design in Artificial Intelligence for...            313   
2     AL803  Software Design for Mobile Apps and Connected ...            350   
3     AL805    Computer Engineering for Network Infrastructure            321   
4     AL810                                 Quantity Surveying            328   
...     ...                                                ...            ...   
1646  SG441                              Environmental Science            NaN   
1647  SG446                                Applied Archaeology            NaN   
1648  TL803                                   Music Technology            NaN   
1649  TL812                       Computing with Digital Media            NaN   
1650  WD148                                  Tourism Marketing            NaN   

     points_r1_2020 points_r1_2019  
0               303            304  
1               332            301  
2               337            309  
3               333            329  
4               319            307  
...             ...            ...  
1646            NaN            297  
1647            NaN            289  
1648            NaN            264  
1649            NaN            369  
1650            NaN            269  

[1651 rows x 5 columns]

## Cleaning up the data


In [ ]:
# Export the data to a CSV file
allcourses.to_csv('All_Courses_Check.csv')